artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/Ubuntu-VM_OperatingSystemInterface-w1/source/src/main/java/org/opentosca/artifacttemplates/ubuntuvm/VirtualMachine.java

```diff
@@ -0,0 +1,200 @@
+                }
+                //noinspection BusyWait
+                Thread.sleep(1000);
+            }
+
+            if (!errorBuffer.toString().isEmpty()) {
+                LOG.error(errorBuffer.toString());
+            }
+            if (!outputBuffer.toString().isEmpty()) {
+                LOG.info(outputBuffer.toString());
+            }
+
+            LOG.info("Command '{}' exited with code '{}' on VM '{}'", command, channel.getExitStatus(), this);
+
+            return outputBuffer.toString().trim();
+        } finally {
+            if (channel != null) {
+                channel.disconnect();
+            }
+        }
+    }
+
+    public void uploadFile(String source, String target) throws Exception {
+        ChannelSftp channel = null;
+        try {
+            LOG.info("Uploading file {} to '{}/{}'", source, this, target);
+
+            channel = (ChannelSftp) session.openChannel("sftp");
+            channel.connect();
+            channel.put(source, target);
+
+            LOG.info("Successfully uploaded file '{}' to '{}/{}'", source, this, target);
+        } finally {
+            if (channel != null) {
+                channel.disconnect();
+            }
+        }
+    }
+
+    public String replaceHome(String command) throws Exception {
+        if (command.contains("~/")) {
+            String pwd = execCommand("pwd").trim();
+
+            String replaced;
+            if (pwd.endsWith("/")) {
+                replaced = command.replaceAll("~/", pwd);
+            } else {
+                replaced = command.replaceAll("~", pwd);
+            }
+
+            LOG.info("Replaced '~' in '{}' with '{}' which results in '{}' on VM '{}'", command, pwd, replaced, this);
+            return replaced;
+        }
+        return command;
+    }
+
+    public void convertToUnix(String target) throws Exception {
+        LOG.info("Converting file '{}' to unix on VM '{}'", target, this);
+
+        if (!target.endsWith(".sh")) {
+            LOG.info("Skipping converting file to unix since file '{}' does not end with .sh", target);
+            return;
+        }
+
+        installPackages("dos2unix");
+        execCommand("dos2unix " + target + " " + target);
+
+        LOG.info("Successfully converted file '{}' to unix on VM '{}'", target, this);
+    }
+
+    public void installPackages(String packages) throws Exception {
+        String command = "(sudo apt-get update && sudo apt-get -y install " + packages + ") || (sudo yum update && sudo yum -y install " + packages + ")";
+        String output = execCommand(command);
+        boolean success = output.endsWith("Complete!") || output.endsWith("Nothing to do");
+        if (!success) throw new Exception("Could not install packages!");
+    }
+
+    private void sleep() {
+        try {
+            Thread.sleep(2500);
+        } catch (InterruptedException e) {
+            LOG.error("Could not sleep on VM '{}'", this, e);
```

Comment:
"sleep on VM" does not sound correct as the IA is sleeping and was interrupted in this case.

Reply:
done